mirrord-layer/src/tcp_mirror.rs

```diff
@@ -0,0 +1,170 @@
+use std::{
+    borrow::Borrow,
+    collections::HashSet,
+    hash::{Hash, Hasher},
+};
+
+use anyhow::Result;
+use async_trait::async_trait;
+use mirrord_protocol::{ConnectionID, NewTCPConnection, TCPClose, TCPData};
+use tokio::{
+    io::{AsyncReadExt, AsyncWriteExt},
+    net::TcpStream,
+    select,
+    sync::mpsc::{channel, Receiver, Sender},
+    task,
+};
+use tokio_stream::{wrappers::ReceiverStream, StreamExt};
+use tracing::{debug, error};
+
+use crate::{common::Listen, tcp::TCPHandler};
+
+async fn tcp_tunnel(mut local_stream: TcpStream, remote_stream: Receiver<Vec<u8>>) {
+    let mut remote_stream = ReceiverStream::new(remote_stream);
+    let mut buffer = vec![0; 1024];
+    loop {
+        select! {
+            message = remote_stream.next() => {
+                match message {
+                    Some(data) => {
+                        match local_stream.write_all(&data).await {
+                            Ok(_) => {},
+                            Err(err) => {error!("writing to local stream err {err:?}"); break;}
+                        }
+                    },
+                    None => {
+                        debug!("tcp tunnel exiting due to remote stream closed");
+                        break;
+                    }
+                }
+            },
+            res = local_stream.read(&mut buffer) => {
+                match res {
+                    Err(ref err) if err.kind() == std::io::ErrorKind::WouldBlock => {
+                        continue;
+                        },
+                    Err(err) => {
+                        error!("local stream received an error reading {err:?}");
+                        break;
+                    }
+                    Ok(n) if n == 0 => {
+                        debug!("tcp tunnel exiting due to local stream closed");
+                        break;
+                    },
+                    Ok(_) => {}
+                }
+
+            }
+        }
+    }
+    debug!("exiting tcp tunnel");
+}
+
+struct Connection {
+    writer: Sender<Vec<u8>>,
+    id: ConnectionID,
+}
+
+impl Eq for Connection {}
+
+impl PartialEq for Connection {
+    fn eq(&self, other: &Connection) -> bool {
+        self.id == other.id
+    }
+}
+
+impl Hash for Connection {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.id.hash(state);
+    }
+}
+
+impl Connection {
+    pub fn new(id: ConnectionID, writer: Sender<Vec<u8>>) -> Self {
+        Self { id, writer }
+    }
+
+    pub async fn write(&mut self, data: Vec<u8>) -> Result<()> {
+        Ok(self.writer.send(data).await?)
+    }
+}
+
+impl Borrow<ConnectionID> for Connection {
+    fn borrow(&self) -> &ConnectionID {
+        &self.id
+    }
+}
+
+/// Handles traffic mirroring
+pub struct TCPMirrorHandler {
+    ports: HashSet<Listen>,
+    running: bool,
```

Comment:
Not a big fan of holding `bool`s to represent On/Off states. Can't you "flip it around" and have some sort of `RunningTCPMirrorHandler` and a `StoppedTCPMirrorHandler` (if the stopped version even makes sense, like why do we care about a non-running thing)?

Reply:
I would suggest turning this struct into a sort of FSM, and implement message handling (just a simple way of handling `event -> next state`, to decouple stateful mutations and have these look more like state transitions instead) to change states between stopped/running/whatever else.

Reply:
There's no use for Running/Stopped, just a way to share behavior between TCPHandler implementations. I can remove that by moving the loop to each implementatino